### ISSUE TITLE
torch.tril and torch.triu

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#12653]
And tril_indices and triu_indices too!

**Comment by vishwakftw** [comment]
Is this an expected implementation?
```
def tril_indices(rows, cols, offset=0):
    return torch.ones(rows, cols, dtype=torch.uint8).tril(offset).nonzero()
```
and analogously for `triu_indices`?

**Comment by zou3519** [comment]
It is, but it might be a good idea to fuse all of those ops together in one if one calls tril_indices a lot. On the other hand this is a one liner so I'm not sure we want this in the API

**Comment by zou3519** [comment]
I thought about how to do this and I think this would be a great first task so I'm reserving it for now. Please do let me know if you wanted to work on it @vishwakftw and I can assign it to you instead

**Comment by vishwakftw** [comment]
I wouldn't mind working on it considering the code is already there above, but if you would like to assign it to someone else, please feel free to :)

**Comment by vishwakftw** [comment]
@zou3519 Shall I send in a PR?

**Comment by SsnL** [comment]
The oneliner is pretty inefficient. So I'd call for a better implementation.

**Comment by vishwakftw** [comment]
Agreeably true. I'm working on batched `triu` and `tril` now. If you want to reserve this for someone else, please do so.

**Comment by SsnL** [comment]
I think @zou3519 wants to reserve this. I'll let him confirm.

**Comment by zou3519** [comment]
Yes, reserving this (`torch.tril_indices`, `torch.triu_indices`), a more efficient version of the one liner is interesting to implement

**Comment by mrshenli** [comment]
Implemented in #14904 and #15203. Shall we close this issue? @SsnL
